spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -303,6 +401,127 @@ object DeltaHistoryManager extends DeltaLogging {
+   */
+  private[delta] def getFirstCommitAndICTAfter(
+      version: Long,
+      upperBoundExclusive: Long,
+      basePath: Path,
+      logStore: LogStore,
+      conf: Configuration,
+      commitFileProvider: DeltaCommitFileProvider): Option[Commit] = {
+    val commitInfoOpt = try {
+      val deltaFile = commitFileProvider.deltaFile(version)
+      getCommitInfoOpt(logStore, deltaFile, conf)
+    } catch {
+      case _: FileNotFoundException => None
+    }
+    if (commitInfoOpt.isDefined) {
+      val timestamp = CommitInfo.getRequiredInCommitTimestamp(commitInfoOpt, version.toString)
+      Some(Commit(version, timestamp))
+    } else {
+      logStore
+        .listFrom(FileNames.listingPrefix(basePath, version), conf)
+        .takeWhile {
+          fs => FileNames.getFileVersionOpt(fs.getPath).forall(_ < upperBoundExclusive)
+        }
+        .collectFirst { case DeltaFile(f, v) =>
+          getCommitWithInCommitTimestamp(v, f, logStore, conf)
+        }
+        .flatten
+    }
+  }
+
+  /**
+   * Returns the latest commit (with its inCommitTimestamp) that happened at or before
+   * `searchTimestamp` in the range `[startCommit.version, end)`.
+   * If no such commit exists, None is returned.
+   *
+   * The algorithm divides the range into `numChunks` chunks. It then finds the last
+   * chunk where the ICT of its first available commit is less than or equal to `searchTimestamp`.
+   * This chunk is then further divided into `numChunks` chunks and the process is repeated.
+   */
+  private[delta] def getActiveCommitAtTimeFromICTRange(
+      searchTimestamp: Long,
+      startCommit: Commit,
+      end: Long,
+      conf: Configuration,
+      basePath: Path,
+      logStore: LogStore,
+      numChunks: Long,
+      spark: SparkSession,
+      commitFileProvider: DeltaCommitFileProvider): Option[Commit] = {
+    require(startCommit.version < end, "start must be less than end")
+    var curStartCommit = startCommit
+    var curEnd = end
+    while (curStartCommit.version < curEnd) {
+      val numVersionsInRange = curEnd - curStartCommit.version
+      val chunkSize = math.max(numVersionsInRange / numChunks, 1)
+
+      // min(chunkSize) = 1 and curStartCommit.version < end
+      // therefore, getChunkEnd(chunkStart) will always be > chunkStart
+      def getChunkEnd(chunkStart: Long): Long = math.min(chunkStart + chunkSize, curEnd)
+
+      val chunkStartICTFutures =
+        (curStartCommit.version until curEnd by chunkSize).map { chunkStart =>
+          threadPool.submit(spark) {
+            Option.when(chunkStart == curStartCommit.version)(curStartCommit)
+              .orElse {
+                getFirstCommitAndICTAfter(
+                  chunkStart,
+                  upperBoundExclusive = getChunkEnd(chunkStart),
+                  basePath,
+                  logStore,
+                  conf,
+                  commitFileProvider
+                )
+              }
```

Comment:
We can use [Future.successful](https://www.scala-lang.org/api/2.12.6/scala/concurrent/Future$.html#successful[T](result:T):scala.concurrent.Future[T]) instead of burning a thread pool entry on chunk starts:
```suggestion
          if (chunkStart == curStartCommit.version) {
            Future.successful(curStartCommit)
          } else {
            threadPool.submit(spark) {
              getFirstCommitAndICTAfter(
                chunkStart,
                upperBoundExclusive = getChunkEnd(chunkStart),
                basePath,
                logStore,
                conf,
                commitFileProvider
              )
            }
          }
```

Reply:
Right, thanks for catching that!